### PR TITLE
Adds a FileConflictStrategy property to CheckoutOptions

### PR DESCRIPTION
**Pull request by sitereactor**
PR for #868 

This is my initial implementation of using a FileConflictStrategy to select Ours/Theirs for Paths upon Checkout. But I'm wondering if it would make _(more)_ sense to make an overload for `CheckoutPaths` that accepts an Ours/Theirs (enum) instead of `CheckoutOptions` with a `FileConflictStrategy` property.
One thing that bugs me about `CheckoutPaths` is that you can pass it a "committishOrBranchSpec" that doesn't correspond to the `FileConflictStrategy` property and you'd thus get an unexpected result (ie. pass the branch spec for the currently checked out branch and `FileConflictStrategy.Theirs`).

Wouldn't it be easier if it was possible to pass a list of `Conflict` objects along with an Ours/Theirs enum to `CheckoutPaths` - or a list of `IndexEntry` objects and an Ours/Theirs enum  - if it is in fact possible to get the correct `Tree` from the `IndexEntry.Id`(?).

@nulltoken whats your thoughts on this?

### CODE REVIEW

**Comment by ethomson** [comment]
I'm concerned about you forcing `Force` here, that seems like it should be unnecessary in the general case - it seems like it should be necessary only when you have a file already in the working directory that you're trying to overwrite, no?  In that case, making a user specify `Force` themselves seems like the safe thing to do.

**Comment by sitereactor** [comment]
@ethomson happy to change it, so it uses the flag set by the user. But would there be cases where checking out paths with Ours/Theirs isn't done in order to resolve a merge conflict where the file exists in both trees - thus needing 'Force' ?

**Comment by ethomson** [comment]
Indeed, I think that there would be cases (many, I would expect, actually).  But I think that the user should be responsible for specifying `Force` in those cases and we should default to `Safe`...  Just my thinking, though.  @nulltoken ?

**Comment by nulltoken** [comment]
> But I think that the user should be responsible for specifying `Force` in those cases and we should default to `Safe`...  Just my thinking, though.  @nulltoken ?

I agree with this

**Comment by nulltoken** [comment]
@sitereactor Does this make sense to you as well?

**Comment by sitereactor** [comment]
@nulltoken yes, im gonna update the PR tomorrow as part of my freedom friday :wink:
If you have additional scenarios that I should add tests cases for I'd be happy to more.

**Comment by sitereactor** [comment]
@nulltoken this PR is now rebased and updated. Let me know if this is more inline with your expectations. I changed the way `Force` is specified as suggested above.

**Comment by nulltoken** [comment]
@ethomson How do you like this?

File: LibGit2Sharp.Tests/CheckoutFixture.cs [hunk 1]
```diff
@@ -280,6 +280,65 @@ public void CanForcefullyCheckoutWithConflictingStagedChanges()
...
         [Fact]
+        public void CanCheckoutWithMergeConflictsSelectingTheirs()
+        {
+            const string firstBranchFileName = "first branch file.txt";
+            const string secondBranchFileName = "second branch file.txt";
+            const string sharedBranchFileName = "first+second branch file.txt";
+            const string firstBranchContent = "The first branches comment";
+            const string secondBranchContent = "The second branches comment";
+
+            string path = SandboxStandardTestRepo();
```
Comment (jamill) [inline LibGit2Sharp.Tests/CheckoutFixture.cs]: Do we need the whole Standard Test Repo, or would this test work with a new empty repo (via `InitNewRepository()`)

**Comment by jamill** [comment]
Is this only meant to be used when checking out a file path? or can it be used when checking out a branch / commit as well?

File: LibGit2Sharp.Tests/CheckoutFixture.cs (hunk 1 above)
Comment (sitereactor) [inline LibGit2Sharp.Tests/CheckoutFixture.cs]: I tried changing it to use `InitNewRepository()` and as long as I start with `PopulateBasicRepository(repo);` it works fine. Without the populate method I get an error `LibGit2Sharp.NotFoundExceptionRevspec 'master' not found`.

**Comment by sitereactor** [comment]
@jamill For now I've only added one test for a scenario that I wanted to use this for. But since its on the `CheckoutOptions` it should work for a regular checkout. I'd be happy to add more test cases.

Do you agree with this scenario:
Branch `branch1` has file `xyz.txt`, and branch `branch2` is currently checked out with file `xyz.txt` changed. Calling `Checkout` with `CheckoutOptions() { FileConflictStrategy = CheckoutFileConflictStrategy.Theirs, CheckoutModifiers = CheckoutModifiers.Force }` should overwrite the changed `xyz.txt` file with the one from `branch1`.

Any other scenarios you'd like to see test cases for please let me know. :wink:

File: LibGit2Sharp/CheckoutOptions.cs [hunk 2]
```diff
@@ -34,8 +39,16 @@ CheckoutStrategy IConvertableToGitCheckoutOpts.CheckoutStrategy
         {
             get
             {
-                return CheckoutModifiers.HasFlag(CheckoutModifiers.Force) ?
-                    CheckoutStrategy.GIT_CHECKOUT_FORCE : CheckoutStrategy.GIT_CHECKOUT_SAFE;
+                var strategy = GitCheckoutOptsWrapper.CheckoutStrategyFromFileConflictStrategy(FileConflictStrategy);
+                if (strategy != default(CheckoutStrategy))
```
Comment (jamill) [inline LibGit2Sharp/CheckoutOptions.cs]: I think this logic could be simplified, maybe something like:

```
var strategy = CheckoutModifiers.HasFlag(CheckoutModifiers.Force) ? CheckoutStrategy.GIT_CHECKOUT_FORCE : CheckoutStrategy.GIT_CHECKOUT_SAFE;
strategy |= GitCheckoutOptsWrapper.CheckoutStrategyFromFileConflictStrategy(FileConflictStrategy);


```

**Comment by jamill** [comment]
> Do you agree with this scenario:

With your example, I don't think need to specify the `FileConflictStrategy` at all - just specifying the `Force` flag will result in the checkout proceeding and the new contents of the changed file to be written. I *think* think the `FileConflictStrategy` option only works when checking out when conflicts already exist in the index.

If this option works in the general checkout case, I think we should add tests covering using it in that case (checking out a ref that results in conflicts, and then verifying the expected state of the index / workdir).

If this does only make sense when the index contains conflicts, then maybe we could either 1) throw if used in an unexpected manner, 2) Not expose this option on `Checkout(...)`, but expose the functionality through a separate function. With a separate function (option 2), we might not have to force the user to specify the `Force` option.

**Comment by nulltoken** [comment]
@sitereactor What's your feeling regarding @jamill's comment?

**Comment by sitereactor** [comment]
@nulltoken as per @jamill's comment I found that the `FileConflictStrategy` is not needed at all when using `repo.CheckoutPaths`. A toggle between the branches will do the trick. So this left me kinda confused as to how the `FileConflictStrategy` should then be used for Checkouts. Maybe my problem is that I'm too focused on the Path. But I think I need some guidance on intent and where to go from here in order to wrap up this pull request.

I'm also wondering if it even makes sense to have the `FileConflictStrategy` available on `CheckoutPaths` (through `CheckoutOptions`) if it has no effect.

**Comment by ethomson** [comment]
> I'm also wondering if it even makes sense to have the FileConflictStrategy available on CheckoutPaths (through CheckoutOptions) if it has no effect.

I'm not sure that I understand this sentence - your test https://github.com/libgit2/libgit2sharp/pull/940/files#diff-0774acc3e67579e1d46230b9f1e28749R323 seems to indicate that you're passing a `FileConflictStrategy.Theirs` to `CheckoutPaths` and getting the correct file from the index checked out.

**Comment by sitereactor** [comment]
Sorry, that was badly formulated. I mean since you can achieve the same by not using the `FileConflictStrategy` with `CheckoutPaths`, and because the branch that is passed to `CheckoutPaths` has to correspond to the choice of `CheckoutFileConflictStrategy.Theirs` or `CheckoutFileConflictStrategy.Ours`.
